packages/api/tests/lib/writeRecords/test-write-granules.js

```diff
@@ -2680,6 +2644,1144 @@ test.serial('writeGranuleFromApi() removes preexisting granule file from postgre
+    })
+  );
+  t.true(await esGranulesClient.exists(completeGranule.granuleId));
+
+  // Create new files for granule. These records will exist in database
+  // during subsequent granule write from message
+  const updatedFiles = [
+    fakeFileFactory({ size: 5 }),
+    fakeFileFactory({ size: 10 }),
+    fakeFileFactory({ size: 15 }),
+    fakeFileFactory({ size: 20 }),
+  ];
+
+  const updatedGranule = {
+    ...completeGranule,
+    files: updatedFiles,
+    status: 'completed',
+  };
+  cumulusMessage.meta.status = 'completed';
+  cumulusMessage.payload.granules[0] = updatedGranule;
+  await writeGranulesFromMessage({
+    cumulusMessage,
+    executionCumulusId,
+    providerCumulusId,
+    knex,
+    granuleModel,
+  });
+
+  const dynamoRecord = await granuleModel.get({ granuleId });
+  const postgresRecord = await granulePgModel.get(knex, {
+    granule_id: granuleId,
+    collection_cumulus_id: collectionCumulusId,
+  });
+  const apiFormattedPostgresGranule
+    = await translatePostgresGranuleToApiGranule({
+      granulePgRecord: postgresRecord,
+      knexOrTransaction: knex,
+    });
+  const esRecord = await esGranulesClient.get(granuleId);
+
+  const expectedGranule = {
+    ...updatedGranule,
+    // should already be part of updatedGranule, but setting here more explicity
+    files: updatedFiles,
+    // These values *must* be set or the message write logic sets them.
+    createdAt: dynamoRecord.createdAt,
+    published: true,
+    timestamp: dynamoRecord.timestamp,
+    updatedAt: dynamoRecord.updatedAt,
+    // Values to be set as they're set by default for all writes (based on message info)
+    // productVolume is also calculated based on payload
+    duration: dynamoRecord.duration,
+    error: {},
+    pdrName: cumulusMessage.payload.pdr.name,
+    productVolume: '50',
+    status: cumulusMessage.meta.status,
+    // These values are set *only* via finder methods in the message, and override
+    // the passed in granule value.   The factory leaves these unset, so they default to zero
+    timeToArchive: 0,
+    timeToPreprocess: 0,
+  };
+
+  // Files array order is not promised to match between datastores
+  [esRecord, dynamoRecord, expectedGranule, apiFormattedPostgresGranule].forEach((record) => {
+    record.files.sort((f1, f2) => sortFilesByBuckets(f1, f2));
+  });
+
+  // Translated postgres granule matches expected updatedGranule
+  // minus model defaults
+  t.deepEqual(
+    apiFormattedPostgresGranule,
+    removeNilProperties(expectedGranule)
+  );
+  t.deepEqual(dynamoRecord, removeNilProperties(expectedGranule));
+  t.deepEqual(omit(esRecord, ['_id']), removeNilProperties(expectedGranule));
+});
+
+test.serial('writeGranulesFromMessage() on update changing granule status to "running", with different files and the same execution, does not update the granule values or files in Postgres/DynamoDB/ES, so the pre-existing values and files will persist', async (t) => {
```

Comment:
Is this a thing that can happen? A message write comes in that changes the granule status to "running" but it has different files and maybe different granule values? Is this like an out-of-order write?

Reply:
So it seems like we don't write files if writeConstraints === true? Or we write them later in the chain? Just trying to understand/remember what the logic is here.

Reply:
Q1: Yes, in speaking with @Jkovarik it seems like something that can happen - whether an out-of-order write or some other reason to write the granule with a new message and/or execution and the status might be changed. I'm not sure I can explain the uses cases that well, as it gets all a bit convoluted, but I did capture all the scenarios that should be tested in the PR description.

Q2: yes, we don't write files if writeConstraints is true. In the write-granules file above, ln 638.

Reply:
So the condition we're trying to sort here is that: 

A granule changing to running with the same execution with modified files should fail to write because you *cannot* update a granule in a completed state (same execution) to running in the message writes (e.g. write constraints set to true). 

What was happening prior was that we'd return the postgres granule from the write granule methods, *then* write apiGranule.files to postgres after *not* writing it to dynamo/es.